plc4go/internal/ads/Interactions.go

```diff
@@ -0,0 +1,261 @@
+				responseChannel <- response
+				close(responseChannel)
+				return nil
+			},
+			func(err error) error {
+				return nil
+			},
+			time.Second); err != nil {
+			close(responseChannel)
+		} else {
+			close(responseChannel)
+		}
+	}()
+	response, err := ReadWithTimeout(responseChannel)
+	if err != nil {
+		return nil, fmt.Errorf("error writing: %v", err)
+	}
+	return response, nil
+}
+
+func (m *Connection) ExecuteAdsAddDeviceNotificationRequest(ctx context.Context, indexGroup uint32, indexOffset uint32, length uint32, transmissionMode model.AdsTransMode, maxDelay uint32, cycleTime uint32) (model.AdsAddDeviceNotificationResponse, error) {
+	responseChannel := make(chan model.AdsAddDeviceNotificationResponse)
+	go func() {
+		request := m.NewAdsAddDeviceNotificationRequest(indexGroup, indexOffset, length, transmissionMode, maxDelay, cycleTime)
+		if err := m.messageCodec.SendRequest(
+			ctx,
+			request,
+			func(message spi.Message) bool {
+				amsTcpPacket, ok := message.(model.AmsTCPPacket)
+				if !ok {
+					return false
+				}
+				return amsTcpPacket.GetUserdata().GetInvokeId() == request.GetUserdata().GetInvokeId()
+			},
+			func(message spi.Message) error {
+				amsTcpPacket := message.(model.AmsTCPPacket)
+				response := amsTcpPacket.GetUserdata().(model.AdsAddDeviceNotificationResponse)
+				responseChannel <- response
+				close(responseChannel)
+				return nil
+			},
+			func(err error) error {
+				return nil
+			},
+			time.Second); err != nil {
+			close(responseChannel)
+		} else {
+			close(responseChannel)
+		}
+	}()
+	response, err := ReadWithTimeout(responseChannel)
+	if err != nil {
+		return nil, fmt.Errorf("error writing: %v", err)
+	}
+	return response, nil
+}
+
+func (m *Connection) ExecuteAdsDeleteDeviceNotificationRequest(ctx context.Context, notificationHandle uint32) (model.AdsDeleteDeviceNotificationResponse, error) {
+	responseChannel := make(chan model.AdsDeleteDeviceNotificationResponse)
+	go func() {
+		request := m.NewAdsDeleteDeviceNotificationRequest(notificationHandle)
+		if err := m.messageCodec.SendRequest(
+			ctx,
+			request,
+			func(message spi.Message) bool {
+				amsTcpPacket, ok := message.(model.AmsTCPPacket)
+				if !ok {
+					return false
+				}
+				return amsTcpPacket.GetUserdata().GetInvokeId() == request.GetUserdata().GetInvokeId()
+			},
+			func(message spi.Message) error {
+				amsTcpPacket := message.(model.AmsTCPPacket)
+				response := amsTcpPacket.GetUserdata().(model.AdsDeleteDeviceNotificationResponse)
+				responseChannel <- response
+				close(responseChannel)
+				return nil
+			},
+			func(err error) error {
+				return nil
+			},
+			time.Second); err != nil {
+			close(responseChannel)
+		} else {
+			close(responseChannel)
+		}
+	}()
+	response, err := ReadWithTimeout(responseChannel)
+	if err != nil {
+		return nil, fmt.Errorf("error writing: %v", err)
+	}
+	return response, nil
+}
+
+func ReadWithTimeout[T spi.Message](ch <-chan T) (T, error) {
```

Comment:
this should get the ctx passed in an create a new ctx from that with timeout. Other than that time.After is leaking timers.